orttraining/orttraining/training_ops/cuda/communication/common.h

```diff
@@ -26,5 +27,79 @@ inline size_t GetAggregatedAlignedAddress(size_t old_addr) {
   return new_addr;
 }
 
+// This function extracts shapes and sizes for tensors indexed by
+// begin, begin + 1, ..., begin + count - 1.
+// tensor_sizes[i]: i-th indexed tensor's size.
+// tensor_shapes[i]: i-th indexed tensor's shape.
+inline void GetTensorShapesAndSizes(
+    const bool is_index_input,                  // Index inputs from the context if true. Otherwise, use outputs.
+    const int begin,                            // Index of the first sent/received tensor in the context.
+    const int count,                            // Number of sent/received tensors.
+    OpKernelContext* ctx,                       // The context.
+    std::vector<size_t>& tensor_sizes,          // tensor_sizes[i] is the size of i-th sent/received tensor in byte.
+    std::vector<TensorShape>& tensor_shapes) {  // tensor_shapes[i] is the i-th sent/received tensor.
+
+  // Helper function to retrieve input or output tensor.
+  auto get_tensor = [&](const int index) -> const Tensor* {
+    if (is_index_input) {
+      return ctx->Input<Tensor>(begin + index);
+    } else {
+      return ctx->Output<Tensor>(begin + index);
+    }
+  };
+
+  // Get tensors and shapes for indexed tensors from context.
+  tensor_sizes.resize(count);
+  tensor_shapes.resize(count);
+  for (int i = 0; i < count; ++i) {
+    const Tensor* tensor = get_tensor(i);
+    tensor_sizes[i] = tensor->SizeInBytes();
+    tensor_shapes[i] = tensor->Shape();
+  }
+}
+
+// Compute shape-related information from given tensor shapes.
+inline void ComputeShapeRelatedInfo(
+    // tensor_shapes[i] is the size of i-th sent/received tensor in byte.
```

Comment:
>tensor_shapes [](start = 7, length = 13)

Nit: typo

Reply:
Ok.

---
In reply to: [459810315](https://github.com/microsoft/onnxruntime/pull/4547#discussion_r459810315) [](ancestors = 459810315)